core/test/mpi/base/bindings.cpp

```diff
@@ -0,0 +1,758 @@
+{
+    auto data = std::vector<TypeParam>{1, 2, 3, 4};
+    auto comm = gko::mpi::communicator::create_world();
+    auto win =
+        gko::mpi::window<TypeParam>(data.data(), 4 * sizeof(TypeParam), comm);
+    ASSERT_NE(win.get(), MPI_WIN_NULL);
+    win.lock_all();
+    win.unlock_all();
+}
+
+
+TYPED_TEST(MpiBindings, CanSendAndRecvValues)
+{
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    auto recv_array = gko::Array<TypeParam>{this->ref};
+    if (my_rank == 0) {
+        auto send_array = std::vector<TypeParam>{1, 2, 3, 4};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::send<TypeParam>(send_array.data(), 4, rank, 40 + rank,
+                                          comm);
+            }
+        }
+    } else {
+        recv_array = gko::Array<TypeParam>{this->ref, 4};
+        gko::mpi::recv<TypeParam>(recv_array.get_data(), 4, 0, 40 + my_rank,
+                                  comm);
+    }
+    if (my_rank != 0) {
+        auto ref_array = gko::Array<TypeParam>{this->ref, {1, 2, 3, 4}};
+        this->assert_equal_arrays(ref_array, recv_array);
+    }
+}
+
+
+TYPED_TEST(MpiBindings, CanNonBlockingSendAndNonBlockingRecvValues)
+{
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    std::vector<TypeParam> send_array;
+    auto recv_array = gko::Array<TypeParam>{this->ref};
+    TypeParam* data;
+    std::vector<MPI_Request> req1;
+    MPI_Request req2;
+    if (my_rank == 0) {
+        send_array = std::vector<TypeParam>{1, 2, 3, 4};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                req1.emplace_back(gko::mpi::i_send<TypeParam>(
+                    send_array.data(), 4, rank, 40 + rank, comm));
+            }
+        }
+    } else {
+        recv_array = gko::Array<TypeParam>{this->ref, 4};
+        req2 = std::move(gko::mpi::i_recv<TypeParam>(recv_array.get_data(), 4,
+                                                     0, 40 + my_rank, comm));
+    }
+    if (my_rank == 0) {
+        auto stat1 = gko::mpi::wait_all(req1);
+    } else {
+        auto stat2 = gko::mpi::wait(req2);
+    }
+    if (my_rank != 0) {
+        auto ref_array = gko::Array<TypeParam>{this->ref, {1, 2, 3, 4}};
+        this->assert_equal_arrays(ref_array, recv_array);
+    }
+}
+
+
+TYPED_TEST(MpiBindings, CanPutValuesWithLockAll)
+{
+    using window = gko::mpi::window<TypeParam>;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    std::vector<TypeParam> data;
+    if (my_rank == 0) {
+        data = std::vector<TypeParam>{1, 2, 3, 4};
+    } else {
+        data = std::vector<TypeParam>{0, 0, 0, 0};
+    }
+    auto win = window(data.data(), 4 * sizeof(TypeParam), comm);
```

Comment:
Interface suggestion: Can't we infer sizeof(TypeParam) via template parameter? Otherwise this leads to different interfaces, some working on the byte level, some on the element level.